### PR TITLE
NAVAND-1038: adopt Data_ref

### PR DESCRIPTION
**Pull request by VysotskiVadim**

### CODE REVIEW

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt [hunk 1]
```diff
@@ -643,3 +648,34 @@ private val fakeDirectionsRoute: DirectionsRoute by lazy {
         .legs(fakeLegs)
         .build()
 }
+
+private fun dataRefToInputStream(directionsResponseJson: DataRef) =
+    InputStreamReader(ByteBufferBackedInputStream(directionsResponseJson.buffer))
+
+private class ByteBufferBackedInputStream(buf: ByteBuffer) : InputStream() {
```
[the remark this answers is not in the document]
Comment (VysotskiVadim) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt]: Nope. I wanted to say that this implementation of input stream relies/uses byte buffer.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt [hunk 2]
```diff
@@ -643,3 +648,36 @@ private val fakeDirectionsRoute: DirectionsRoute by lazy {
         .legs(fakeLegs)
         .build()
 }
+
+private fun DataRef.asReader(): Reader =
+    InputStreamReader(ByteBufferBackedInputStream(buffer))
+
+private class ByteBufferBackedInputStream(
```
Comment (dzinad) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:655]: What's "backed"? Does it mean "backwards"?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/SDKRouteParser.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    fun parseDirectionsResponse(
+        response: DataRef,
+        request: String,
+        routerOrigin: RouterOrigin,
+    ): Expected<String, List<RouteInterface>>
+
+    companion object {
+        var default: SDKRouteParser = NativeRouteParserWrapper()
+            @VisibleForTesting set
```
Comment (dzinad) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/SDKRouteParser.kt:26]: Why do you need to make it var with visible-for-testing setter? You can simply mock the object.
Then you can make it val, remove unused setter and the `resetDefault` method.

Comment (VysotskiVadim) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/SDKRouteParser.kt:26]: I'm trying to avoid performance penalty static mocking by mockk brings.

Comment (dzinad) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/SDKRouteParser.kt:26]: Static mocking is only done in tests while unused methods stay in our production code, increase aar size, etc.

**Review by DzmitryFomchyn — APPROVED** [review]

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (VysotskiVadim) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:655]: I don't test it directly, but we have tests that create routes through the data ref. Do you see any scenarios rather than serialise a few routes using this input stream to add?

[the remark this answers is not in the document]
Comment (VysotskiVadim) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:655]: However we don't check all the property of that routes. Other tests won't fail if one of the property won't be serialised. I will add a test that compares all the properties

[the remark this answers is not in the document]
Comment (VysotskiVadim) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:655]: Added `navigation routes serialised from string, data_ref, and model are equals` test

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/SDKRouteParser.kt (hunk 3 above)
Comment (VysotskiVadim) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/SDKRouteParser.kt]: @dzinad , @DzmitryFomchyn , it seems that static mocking here slows UI unit tests for ~2 minutes. I see numbers only from 2 runs, so maybe it's just a misfortunate, but still.

Comment (VysotskiVadim) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/SDKRouteParser.kt]: however I don't mind using mocking here, 20 or 22 minutes isn't a big difference 🙂

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (dzinad) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:655]: We can also make this class non-private and test it directly with corner cases (empty buffer, we have to read multiple times, the last read is partial, etc.).

[the remark this answers is not in the document]
Comment (dzinad) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:655]: Not a blocker, approving.

**Review by dzinad — APPROVED** [review]
